src/section/mod.rs

```diff
@@ -352,12 +346,37 @@ impl Section {
     }
 }
 
+/// Error when creating section.
+#[allow(missing_docs)]
+#[derive(Debug, Error)]
+pub enum CreateError {
+    #[error("elders info signing key is not in the chain")]
+    UntrustedEldersInfoSigningKey,
+    #[error("failed to serialize data for signing: {}", .0)]
+    Serialize(#[from] bincode::Error),
+    #[error("failed to combine signature shares: {}", .0)]
+    // TODO: add the `#[from]` attribute here once threshold_crypto publishes the version where
+    // their `Error` implements `std::error::Error`.
+    CombineSignatures(bls::error::Error),
+}
+
+/// Error when merging sections.
+#[derive(Debug, Error)]
```

Comment:
MergeError doesn't require `#[allow(missing_docs)]` like others ?

Reply:
No, because it's not exposed outside of the crate. Hm, but the other types aren't exposed either, so maybe I can remove the attribute there too...

Reply:
also maybe consider make them `pub(crate)` ?

Reply:
That won't work because those types are part of `Error` which is itself exposed. This is a bit confusing but it seems that in rust there is a difference between something being public and it being exposed. You can have things that are `pub` but still not exposed. That means the users of the crate can still use those types, but for example calling methods on instances of them, but they can't construct or deconstruct (`match`) them. 

So in this case we have the `Error` type which is exposed, but some of its variants contain types that are not. So the users of routing can do something like:

```rust
match error {
    Error::InvalidMessage(inner_error) => println!("invalid message: {}", inner_error); // you can use `inner_error` even though you don't know its type.
    // ...
}
```

but they can't do this:

```rust
match error {
    Error::InvalidMessage(inner_error) => {
        match inner_error {
            IntegrityError::Untrusted => ...,
            // ..
        }
    }
    // ...
}
```

because the `IntegrityError` type is not exposed. 

I'm not sure this is the best thing to do. Maybe we should expose those inner errors as well? The reason I decided to to it this way is that those inner errors are mostly interesting for routing itself and not that much for its users. But not sure.

Reply:
I reckon if any part of an Error is in the API then the whole thing needs to be pub visible. I am much more a fan of pub(crate) for readability TBH, otherwise, we need all devs to know the backflips ;-) 

If any of these sub errors do expose then we need to make certain they are in a module that is very well contained. Reason being we can have boundaries between modules as you would expect but we don't want Error types to create another set of boundaries, they need to be the same.

Reply:
Right, so then we have these options:

1. Do not use the inner errors at all. So instead of `Error::InvalidMessage(IntegrityError)` how just `Error::InvalidMessage`.
2. Use type-erased inner errors. So `Error::InvalidMessage(Box<dyn std::error::Error>)` or even `Error::InvalidMessage(String)`
3. Use concrete inner errors, but don't expose them (this is what's in the PR now)
4. Use concrete inner errors and expose them

I don't really like option 1, because that's basically what we have now and I don't think the information in the error is sufficient. As for option 4, this would be my preference, but it would be also this "mini errors" thing which you said you don't want? That leaves us with options 2 or 3. I don't have a strong preference among them as they are effectively very similar. 

Btw, we can't use `pub(crate)` on the inner errors as that doesn't compile.

Reply:
Can we not have InvalidMessage with a `From` clause that is IntegrityError where IntegrityError has more info if required? 

I am cool with mini Errors iff they are module specific and we can test that module independently etc. It's a good look, my issue is mini errors that are not also code modules.

Reply:
> Can we not have InvalidMessage with a From clause that is IntegrityError where IntegrityError has more info if required?

Do you mean `impl From<IntegrityError> for Error` which returns a `Error::InvalidMessage` ? Because we already have that. The question is whether to also expose `IntegrityError` as part of the public API.

Reply:
OK, I exposed the inner errors and updated the PR. I also made them all `#[non_exhaustive]`.